p2panda-rs/src/encoder.rs

```diff
@@ -0,0 +1,286 @@
+use std::convert::{TryFrom, TryInto};
+
+use bamboo_rs_core::entry::{decode, MAX_ENTRY_SIZE};
+use bamboo_rs_core::{Entry as BambooEntry, Signature as BambooSignature, YamfHash};
+
+use crate::atomic::error::EntrySignedError;
+use crate::atomic::Blake2BArrayVec;
+use crate::atomic::{Entry, EntrySigned, Hash, LogId, Message, MessageEncoded, SeqNum};
+use crate::key_pair::KeyPair;
+use arrayvec::ArrayVec;
+use ed25519_dalek::PublicKey;
+
+/// Takes an [`EntrySigned`] and a [`MessageEncoded`], validates the encoded message against the
+/// entry payload hash, returns the decoded message in the form of a [`Message`] if valid otherwise
+/// throws an error.
+pub fn validate_message(
+    entry_encoded: &EntrySigned,
+    message_encoded: &MessageEncoded,
+) -> Result<Message, EntrySignedError> {
+    // Convert to Entry from bamboo_rs_core first
+    let entry: BambooEntry<ArrayVec<[u8; 64]>, ArrayVec<[u8; 64]>> = entry_encoded.into();
+    // Messages may be omitted because the entry still contains the message hash. If the
+    // message is explicitly included we require its hash to match.
+    let message = match message_encoded {
+        msg => {
+            let yamf_hash: YamfHash<Blake2BArrayVec> = (&msg.hash()).to_owned().into();
+
+            if yamf_hash != entry.payload_hash {
+                return Err(EntrySignedError::MessageHashMismatch);
+            }
+            Message::from(msg)
+        }
+    };
+    Ok(message)
+}
```

Comment:
We could call this `safe_decode_message`. Decoding seems to be the primary function with the addition that the hash is validated.

Is this meant to be public? It's only used in this module.

The match seems to duplicate what's happening in `decode_entry`. Would it be better to let this function reject an empty message so that it can only be used when you already know that a message exists?

Reply:
Yeh, you're right, there is some duplication going on in `decode_entry`. Right now that's only where it's used, but there was talk of it being useful in `aquadoggo`, mainly for validating messages though I think. Maybe the validation method should be moved (back??) into `EntrySigned` as a signed entry is always needed for the validation anyway. Then I don't think we even need this method...